### PR TITLE
[SMALLFIX] Use CountDownLatch instead of sleeping in a loop

### PR DESCRIPTION
**Pull request by aaudiber**
Followup to https://github.com/amplab/tachyon/pull/2218

### CODE REVIEW

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 1]
```diff
@@ -71,14 +71,14 @@
   private AMRMClientAsync<ContainerRequest> mRMClient;
   /** Client to talk to Node Manager */
   private NMClient mNMClient;
-  /** Whether a container for Tachyon master is allocated */
-  private volatile boolean mMasterContainerAllocated;
-  /** Num of allocated worker containers */
-  private volatile AtomicInteger mNumAllocatedWorkerContainers;
+  /** The count starts at 1, then becomes 0 when we allocate a container for the Tachyon master */
+  private CountDownLatch mMasterContainerAllocated;
```
Comment (calvinjia) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: Is this kind of abusing `CountDownLatch`?

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 2]
```diff
@@ -71,14 +71,14 @@
...
+  /** Latch counting down the number of workers to allocate before all mNumWorkers are allocated */
+  private CountDownLatch mAllWorkersAllocatedLatch;
```
Comment (calvinjia) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: the naming is inconsistent

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 3]
```diff
@@ -278,7 +270,7 @@ private void launchTachyonWorkerContainers(List<Container> containers) {
         FormatUtils.getSizeFromBytes((long) mRamdiskMemInMB * Constants.MB));
 
     for (Container container : containers) {
-      if (mNumAllocatedWorkerContainers.get() >= mNumWorkers) {
+      if (mAllWorkersAllocatedLatch.getCount() <= 0) {
```
Comment (calvinjia) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: `== 0` for consistency?

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java (hunk 1 above)
Comment (aaudiber) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: I think it's the best tool for the job, but I'm open to other suggestions. The docs for `CountDownLatch` say 

> A CountDownLatch is a versatile synchronization tool and can be used for a number of purposes. A CountDownLatch initialized with a count of one serves as a simple on/off latch, or gate: all threads invoking await wait at the gate until it is opened by a thread invoking countDown()

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java (hunk 2 above)
Comment (aaudiber) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: fixed

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java (hunk 3 above)
Comment (aaudiber) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: changed

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 4]
```diff
@@ -71,14 +71,14 @@
   private AMRMClientAsync<ContainerRequest> mRMClient;
   /** Client to talk to Node Manager */
   private NMClient mNMClient;
-  /** Whether a container for Tachyon master is allocated */
-  private volatile boolean mMasterContainerAllocated;
-  /** Num of allocated worker containers */
-  private volatile AtomicInteger mNumAllocatedWorkerContainers;
+  /** The count starts at 1, then becomes 0 when we allocate a container for the Tachyon master */
+  private CountDownLatch mMasterContainerAllocatedLatch;
```
Comment (yupeng9) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: final?

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java [hunk 5]
```diff
@@ -71,14 +71,14 @@
...
+  /** Latch counting down the number of workers to allocate before all mNumWorkers are allocated */
+  private CountDownLatch mAllWorkersAllocatedLatch;
```
Comment (yupeng9) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: final?

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java (hunk 4 above)
Comment (aaudiber) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: Added, thanks

File: integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java (hunk 5 above)
Comment (aaudiber) [inline integration/yarn/src/main/java/tachyon/yarn/ApplicationMaster.java]: Added, thanks

**Comment by yupeng9** [comment]
LGTM

**Pull request merged**